2018/001-containers-tagging-strategy.md

```diff
@@ -0,0 +1,189 @@
+added to a cluster that is running silently with outdated images. This is
+particularly tricky because it can easily end up in a situation where the
+cluster is running multiple versions of the same image. Let's imagine the
+cluster was running some outdated mariadb image, then when a new node is added
+and this node pulls the images from the registry (including the mariadb one)
+it will pull and run different versions of the images without noticing it.
+Running a cluster with multiple mariadb versions without control or being
+enforced by the administrator is clearly dangerous and painful to debug if it
+causes some issue at some point.
+
+The tagging strategy of the delivered images must be aligned with the build and
+update workflow. Otherwise it is pretty simple to end up in similar situations
+as exposed above.
+
+## Current situation
+
+Before SUSE container registry images have been and are being delivered wrapped
+in RPMs. This way the update and delivery worklfow is the same as any other
+packages. However within the containers ecosystem this presents some isssues:
+
+* Containers ecosystems based on Kubernetes are not disigned to interact with
+  RPM repositories but container registries.
+  
+* Installing, updating and uninstalling RPMs can be costly. In CaaSP it
+  requires a transactional update.
+  
+* Using RPMS requires additional services to handle images upload into docker
+  or crio daemons. See container-feeder tool or sle2docker. While systems like
+  Kubernetes are already prepared to pull and load images from a registry.
+  
+* Big downloads, RPMs do not benefit from layers reuse. RPMs contain all
+  container layers even only unique layers are finally loaded into the daemon. 
+  
+In this context stable tags for references are being used. For CaaSP images are
+tagged with a fixed tag (usually the version of the main containerized
+application) defined in the kiwi file description and not modified during
+updates, however the RPM includes a metadata file used by the container-feeder.
+At loading time container-feeder re-tags the image with the '<tag>-<release>'
+and 'latest' tags, this way, even using stable tags in manifests the images
+are locally tagged with multiple tags, including the specific release tight to
+an RPM and image build.
+
+## Proposed change
+
+Introduce an app or service to track updates in the registry, something similar
+as what zypper does for RPMs and repositories. This is a key point and the
+proposed strategy here assumes there is going to be something like that
+available.
+
+The proposed change mimics the tagging strategy exposed above when using images
+wrapped in RPMs, but using the SUSE container registry and OBS instead.
+
+`kiwi` and `skopeo` tools can build images with multiple tags (since versions
+9.15.3 and 1.30.0 respectively). Thus the main idea is to keep using stable
+tags to facilitate image references in manifests but also tag images in
```

Comment:
> Thus the main idea is to keep using stable tags to facilitate image references in manifests

Currently, we don't reference "stable" tags - we reference the most specific tag available (i.e. the one that includes the build service's build number). Otherwise, we would have hit the problem you described above :)

Reply:
Oh I was not aware that there was a manifest rebuild for each image rebuild, I did not see this part of the chain...